src/js/components/AppListItemComponent.jsx

```diff
@@ -19,6 +20,94 @@ var AppListItemComponent = React.createClass({
     model: React.PropTypes.object.isRequired
   },
 
+  getInitialState: function () {
+    return {
+      numberOfVisibleLabels: -1
+    };
+  },
+
+  componentDidMount: function () {
+    // Avoid referencing window from Node context
+    if (global.window != null) {
+      window.addEventListener("resize", this.handleResize);
+      window.addEventListener("focus", this.handleResize);
+    }
+    this.updateNumberOfVisibleLabels();
+
+  },
+
+  componentDidUpdate: function (prevProps) {
+    if (this.didPropsChange(prevProps)) {
+      this.updateNumberOfVisibleLabels();
+    }
+  },
+
+  componentWillUnmount: function () {
+    if (global.window != null) {
+      window.removeEventListener("resize", this.handleResize);
+      window.removeEventListener("focus", this.handleResize);
+    }
+  },
+
+  shouldComponentUpdate: function (nextProps, nextState) {
+    if (nextState.numberOfVisibleLabels === -1 && !this.props.model.isGroup) {
+      this.updateNumberOfVisibleLabels();
+      return false;
+    }
+
+    return this.didPropsChange(nextProps) ||
+      this.state.numberOfVisibleLabels !== nextState.numberOfVisibleLabels;
+  },
+
+  didPropsChange: function (newProps) {
+    var model = this.props.model;
+    var newModel = newProps.model;
+
+    return model.status !== newModel.status ||
+      model.tasksRunning !== newModel.tasksRunning ||
+      !Util.isArrayEqual(model.health, newModel.health) ||
+      model.totalMem !== newModel.totalMem ||
+      model.totalCpus !== newModel.totalCpus ||
+      model.instances !== newModel.instances ||
+      !Util.isArrayEqual(Object.keys(model.labels),
+        Object.keys(newModel.labels));
```

Comment:
@orlandohohmeier  @aldipower 
heads up: this check is wrong and produces an Error in the console.
It is wrong because it only checks the label keys, but not the values, and it throws an error when an object has no labels. 

I am not too happy about this block anyway, not so readable.